File: src/hideout/chunk.py
from __future__ import annotations

import hashlib
import re
from dataclasses import asdict, dataclass
from pathlib import Path

from hideout.config import get_config

HEADING_RE = re.compile(r"^(#{1,6})\s+(.*)$")
PAGE_RE = re.compile(
    r"^\*PDF page (\d+)(?: \(printed (\d+)\))?\*$"
)
SKIP_FILES = {"README.md"}


@dataclass
class Chunk:
    id: str
    file: str
    book: str
    chapter: str
    headings: list[str]
    page: int | None
    printed: int | None
    text: str
    source_mtime: float = 0.0

    def embed_text(self) -> str:
        path = " > ".join([self.chapter, *self.headings]) if self.headings else self.chapter
        page = f"PDF page {self.page}" if self.page else ""
        if self.printed is not None:
            page += f" (printed {self.printed})"
        header = "\n".join(p for p in (self.book, path, page) if p)
        return f"{header}\n\n{self.text}"

    def as_dict(self) -> dict:
        return asdict(self)


def source_roots() -> list[Path]:
    return [p for p in get_config().sources if p.is_dir()]


def markdown_files(roots: list[Path] | None = None) -> list[tuple[Path, Path]]:
    roots = roots if roots is not None else source_roots()
    out: list[tuple[Path, Path]] = []
    for root in roots:
        for path in root.rglob("*.md"):
            if path.name in SKIP_FILES:
                continue
            out.append((path, root))
    return sorted(out, key=lambda item: (str(item[1]), str(item[0])))


def source_hash(files: list[tuple[Path, Path]] | None = None) -> str:
    items = files if files is not None else markdown_files()
    h = hashlib.sha256()
    for path, root in items:
        rel = path.relative_to(root)
        h.update(str(root.resolve()).encode())
        h.update(str(rel).encode())
        h.update(path.read_bytes())
    return h.hexdigest()[:16]


def _book_and_file(path: Path, root: Path, *, prefix: bool) -> tuple[str, str]:
    rel = path.relative_to(root)
    book = rel.parts[0] if len(rel.parts) > 1 else root.name
    rel_posix = rel.as_posix()
    if prefix:
        return f"{root.name}/{book}", f"{root.name}/{rel_posix}"
    return book, rel_posix


def _flush(
    chunks: list[Chunk],
    file: Path,
    root: Path,
    chapter: str,
    headings: list[str],
    page: int | None,
    printed: int | None,
    buf: list[str],
    mtime: float,
    *,
    prefix: bool,
) -> None:
    cfg = get_config()
    text = re.sub(r"\n{3,}", "\n\n", "\n".join(buf)).strip()
    if len(text) < cfg.min_chars:
        return
    parts = _split_text(text) if len(text) > cfg.max_chars else [text]
    base = len(chunks)
    book, rel = _book_and_file(file, root, prefix=prefix)
    root_key = str(root.resolve())
    for i, part in enumerate(parts):
        slug = _slug("-".join(headings) or chapter)
        chunk_id = f"{root_key}:{rel}:{slug}:{base + i}"
        chunks.append(
            Chunk(
                id=chunk_id,
                file=rel,
                book=book,
                chapter=chapter,
                headings=list(headings),
                page=page,
                printed=printed,
                text=part,
                source_mtime=mtime,
            )
        )


def _split_text(text: str) -> list[str]:
    cfg = get_config()
    paras = [p.strip() for p in re.split(r"\n\s*\n", text) if p.strip()]
    out: list[str] = []
    buf: list[str] = []
    size = 0
    for para in paras:
        extra = len(para) + 2
        if buf and size + extra > cfg.max_chars:
            out.append("\n\n".join(buf))
            overlap: list[str] = []
            osize = 0
            for prev in reversed(buf):
                if osize + len(prev) > cfg.overlap_chars:
                    break
                overlap.append(prev)
                osize += len(prev)
            buf = list(reversed(overlap))
            size = sum(len(x) + 2 for x in buf)
        buf.append(para)
        size += extra
    if buf:
        out.append("\n\n".join(buf))
    return out or [text[: cfg.max_chars]]


def _slug(text: str) -> str:
    text = text.lower().replace("ä", "a").replace("ö", "o").replace("å", "a")
    text = re.sub(r"[^a-z0-9]+", "-", text).strip("-")
    return text[:80] or "section"


def _heading_stack(stack: list[tuple[int, str]], level: int, title: str) -> list[tuple[int, str]]:
    stack = [item for item in stack if item[0] < level]
    stack.append((level, title.strip()))
    return stack


def chunk_file(path: Path, root: Path, *, prefix: bool = False) -> list[Chunk]:
    chapter = path.stem
    headings: list[tuple[int, str]] = []
    page: int | None = None
    printed: int | None = None
    chunk_page = page
    chunk_printed = printed
    buf: list[str] = []
    chunks: list[Chunk] = []
    mtime = path.stat().st_mtime

    def flush() -> None:
        nonlocal buf
        _flush(
            chunks,
            path,
            root,
            chapter,
            [title for level, title in headings if level != 1],
            chunk_page,
            chunk_printed,
            buf,
            mtime,
            prefix=prefix,
        )
        buf = []

    for raw in path.read_text(encoding="utf-8").splitlines():
        line = raw.rstrip()
        page_match = PAGE_RE.match(line)
        if page_match:
            page = int(page_match.group(1))
            printed = int(page_match.group(2)) if page_match.group(2) else None
            if not buf:
                chunk_page, chunk_printed = page, printed
            continue
        heading = HEADING_RE.match(line)
        if heading:
            flush()
            level = len(heading.group(1))
            title = heading.group(2).strip()
            headings = _heading_stack(headings, level, title)
            if level == 1:
                chapter = title
            chunk_page, chunk_printed = page, printed
            continue
        if line.strip():
            buf.append(line)
        elif buf:
            buf.append("")

    flush()
    return chunks


def chunk_all() -> list[Chunk]:
    files = markdown_files()
    roots = {root for _, root in files}
    prefix = len(roots) > 1
    chunks: list[Chunk] = []
    for path, root in files:
        chunks.extend(chunk_file(path, root, prefix=prefix))
    return chunks

File: src/chunk.ts
import { createHash } from "node:crypto";
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { basename, join, relative, resolve, sep } from "node:path";
import { getConfig } from "./config.js";

const HEADING_RE = /^(#{1,6})\s+(.*)$/;
const PAGE_RE = /^\*PDF page (\d+)(?: \(printed (\d+)\))?\*$/;
const SKIP_FILES = new Set(["README.md"]);

export type Chunk = {
  id: string;
  file: string;
  book: string;
  chapter: string;
  headings: string[];
  page: number | null;
  printed: number | null;
  text: string;
  sourceMtime: number;
};

export function embedText(chunk: Chunk): string {
  const path = chunk.headings.length
    ? [chunk.chapter, ...chunk.headings].join(" > ")
    : chunk.chapter;
  let page = chunk.page ? `PDF page ${chunk.page}` : "";
  if (chunk.printed != null) page += ` (printed ${chunk.printed})`;
  const header = [chunk.book, path, page].filter(Boolean).join("\n");
  return `${header}\n\n${chunk.text}`;
}

export function sourceRoots(): string[] {
  return getConfig().sources.filter((p) => {
    try {
      return existsSync(p) && statSync(p).isDirectory();
    } catch {
      return false;
    }
  });
}

function walkMarkdown(root: string): string[] {
  const out: string[] = [];
  const walk = (dir: string) => {
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const path = join(dir, entry.name);
      if (entry.isDirectory()) walk(path);
      else if (entry.isFile() && entry.name.endsWith(".md") && !SKIP_FILES.has(entry.name)) {
        out.push(path);
      }
    }
  };
  walk(root);
  return out;
}

export function markdownFiles(roots?: string[]): [string, string][] {
  const dirs = roots ?? sourceRoots();
  const out: [string, string][] = [];
  for (const root of dirs) {
    for (const path of walkMarkdown(root)) {
      out.push([path, root]);
    }
  }
  return out.sort((a, b) => {
    const rootCmp = a[1].localeCompare(b[1]);
    return rootCmp !== 0 ? rootCmp : a[0].localeCompare(b[0]);
  });
}

export function sourceHash(files?: [string, string][]): string {
  const items = files ?? markdownFiles();
  const h = createHash("sha256");
  for (const [path, root] of items) {
    const rel = relative(root, path);
    h.update(resolve(root));
    h.update(rel);
    h.update(readFileSync(path));
  }
  return h.digest("hex").slice(0, 16);
}

function asPosix(path: string): string {
  return path.split(sep).join("/");
}

function bookAndFile(path: string, root: string, prefix: boolean): [string, string] {
  const rel = asPosix(relative(root, path));
  const parts = rel.split("/");
  const book = parts.length > 1 ? parts[0] : basename(root);
  if (prefix) return [`${basename(root)}/${book}`, `${basename(root)}/${rel}`];
  return [book, rel];
}

function slug(text: string): string {
  const cleaned = text
    .toLowerCase()
    .replaceAll("ä", "a")
    .replaceAll("ö", "o")
    .replaceAll("å", "a")
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
  return cleaned.slice(0, 80) || "section";
}

function splitText(text: string): string[] {
  const cfg = getConfig();
  const paras = text
    .split(/\n\s*\n/)
    .map((p) => p.trim())
    .filter(Boolean);
  const out: string[] = [];
  let buf: string[] = [];
  let size = 0;
  for (const para of paras) {
    const extra = para.length + 2;
    if (buf.length && size + extra > cfg.maxChars) {
      out.push(buf.join("\n\n"));
      const overlap: string[] = [];
      let osize = 0;
      for (let i = buf.length - 1; i >= 0; i--) {
        if (osize + buf[i].length > cfg.overlapChars) break;
        overlap.push(buf[i]);
        osize += buf[i].length;
      }
      buf = overlap.reverse();
      size = buf.reduce((n, x) => n + x.length + 2, 0);
    }
    buf.push(para);
    size += extra;
  }
  if (buf.length) out.push(buf.join("\n\n"));
  return out.length ? out : [text.slice(0, cfg.maxChars)];
}

function flush(
  chunks: Chunk[],
  file: string,
  root: string,
  chapter: string,
  headings: string[],
  page: number | null,
  printed: number | null,
  buf: string[],
  mtime: number,
  prefix: boolean,
): void {
  const cfg = getConfig();
  const text = buf.join("\n").replace(/\n{3,}/g, "\n\n").trim();
  if (text.length < cfg.minChars) return;
  const parts = text.length > cfg.maxChars ? splitText(text) : [text];
  const base = chunks.length;
  const [book, rel] = bookAndFile(file, root, prefix);
  const rootKey = resolve(root);
  for (let i = 0; i < parts.length; i++) {
    const headingSlug = slug(headings.join("-") || chapter);
    chunks.push({
      id: `${rootKey}:${rel}:${headingSlug}:${base + i}`,
      file: rel,
      book,
      chapter,
      headings: [...headings],
      page,
      printed,
      text: parts[i],
      sourceMtime: mtime,
    });
  }
}

function headingStack(
  stack: [number, string][],
  level: number,
  title: string,
): [number, string][] {
  const next = stack.filter(([lvl]) => lvl < level);
  next.push([level, title.trim()]);
  return next;
}

export function chunkFile(path: string, root: string, prefix = false): Chunk[] {
  let chapter = basename(path, ".md");
  let headings: [number, string][] = [];
  let page: number | null = null;
  let printed: number | null = null;
  let chunkPage: number | null = page;
  let chunkPrinted: number | null = printed;
  let buf: string[] = [];
  const chunks: Chunk[] = [];
  const mtime = statSync(path).mtimeMs / 1000;

  const doFlush = () => {
    flush(
      chunks,
      path,
      root,
      chapter,
      headings.filter(([level]) => level !== 1).map(([, title]) => title),
      chunkPage,
      chunkPrinted,
      buf,
      mtime,
      prefix,
    );
    buf = [];
  };

  for (const raw of readFileSync(path, "utf8").split(/\r?\n/)) {
    const line = raw.replace(/\s+$/, "");
    const pageMatch = PAGE_RE.exec(line);
    if (pageMatch) {
      page = Number(pageMatch[1]);
      printed = pageMatch[2] ? Number(pageMatch[2]) : null;
      if (!buf.length) {
        chunkPage = page;
        chunkPrinted = printed;
      }
      continue;
    }
    const heading = HEADING_RE.exec(line);
    if (heading) {
      doFlush();
      const level = heading[1].length;
      const title = heading[2].trim();
      headings = headingStack(headings, level, title);
      if (level === 1) chapter = title;
      chunkPage = page;
      chunkPrinted = printed;
      continue;
    }
    if (line.trim()) buf.push(line);
    else if (buf.length) buf.push("");
  }
  doFlush();
  return chunks;
}

export function chunkAll(): Chunk[] {
  const files = markdownFiles();
  const roots = new Set(files.map(([, root]) => root));
  const prefix = roots.size > 1;
  const chunks: Chunk[] = [];
  for (const [path, root] of files) {
    chunks.push(...chunkFile(path, root, prefix));
  }
  return chunks;
}
